tests/Microsoft.Identity.Test.Unit/MetricsTests.cs

```diff
@@ -0,0 +1,128 @@
+                Assert.AreEqual(0, Metrics.TotalAccessTokensFromCache);
+
+                // Act - AcquireTokenForClient returns result from cache
+                result = await cca.AcquireTokenForClient(TestConstants.s_scope.ToArray()).ExecuteAsync(CancellationToken.None).ConfigureAwait(false);
+
+                Assert.IsNotNull(result);
+                Assert.AreEqual(TokenSource.Cache, result.AuthenticationResultMetadata.TokenSource);
+                Assert.IsTrue(result.AuthenticationResultMetadata.DurationInCacheInMs > 0);
+                Assert.IsTrue(result.AuthenticationResultMetadata.DurationInHttpInMs == 0);
+                Assert.IsTrue(result.AuthenticationResultMetadata.DurationTotalInMs > 0);
+                Assert.AreEqual(1, Metrics.TotalAccessTokensFromIdP);
+                Assert.AreEqual(1, Metrics.TotalAccessTokensFromCache);
+                Assert.IsTrue(Metrics.TotalDurationInMs > 0);
+            }
+        }
+
+        [TestMethod]
+        public async Task MetricsUpdatedSucessfully_AcquireTokenInteractive_Async()
+        {
+            using (var harness = CreateTestHarness())
+            {
+                harness.HttpManager.AddInstanceDiscoveryMockHandler();
+                harness.HttpManager.AddSuccessTokenResponseMockHandlerForPost();
+
+                PublicClientApplication pca = PublicClientApplicationBuilder.Create(TestConstants.ClientId)
+                                            .WithAuthority(new Uri(ClientApplicationBase.DefaultAuthority), false)
+                                            .WithHttpManager(harness.HttpManager)
+                                            .BuildConcrete();
+
+                InMemoryTokenCache memoryTokenCache = new InMemoryTokenCache(withOperationDelay: true, shouldClearExistingCache: false);
+                memoryTokenCache.Bind(pca.UserTokenCache);
+
+                pca.ServiceBundle.ConfigureMockWebUI(
+                    AuthorizationResult.FromUri(pca.AppConfig.RedirectUri + "?code=some-code"));
+
+                var result = await pca
+                    .AcquireTokenInteractive(TestConstants.s_scope)
+                    .ExecuteAsync(CancellationToken.None).ConfigureAwait(false);
+
+                Assert.IsNotNull(result);
+                Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
+                Assert.IsTrue(result.AuthenticationResultMetadata.DurationInCacheInMs > 0);
+                Assert.IsTrue(result.AuthenticationResultMetadata.DurationTotalInMs > 0);
+                Assert.AreEqual(1, Metrics.TotalAccessTokensFromIdP);
+                Assert.AreEqual(0, Metrics.TotalAccessTokensFromCache);
+                Assert.IsTrue(Metrics.TotalDurationInMs > 0);
+            }
+        }
+
+        [TestMethod]
+        public async Task MetricsUpdatedSucessfully_AcquireTokenSilent_Async()
+        {
+            using (var harness = CreateTestHarness())
```

Comment:
nit: move silent w/interactive for more interesting results maybe?
can you add the metrics to other tests, like ones that are testing the TokenSource, you could also check the metrics there as well. just a thought.

Reply:
#Resolved